src/main/java/com/facebook/openwifirrm/modules/ModelerUtils.java

```diff
@@ -202,4 +209,89 @@ public static double calculateTPCMetrics(
+		return entry1.bssid.equals(entry2.bssid) && Objects.equals(entry1.ssid, entry2.ssid)
+				&& entry1.frequency == entry2.frequency && entry1.channel == entry2.channel
+				&& Objects.equals(entry1.ht_oper, entry2.ht_oper) && Objects.equals(entry1.vht_oper, entry2.vht_oper);
+	}
+
+	/**
+	 *
+	 * @param dataModel
+	 * @param obsoletionPeriodMs per-BSSID, the maximum amount of time (in
+	 *                           milliseconds) it is worth aggregating over,
+	 *                           starting from the most recent scan entry and
+	 *                           working backwards in time on a per-BSSID basis
+	 *                           (per-BSSID basis meaning that this method takes the
+	 *                           most recent scan entry for a specific BSSID and
+	 *                           looks back from there and repeats this for every
+	 *                           BSSID). An entry exactly {@code obsoletionPeriodMs}
+	 *                           ms earlier than the most recent entry is considered
+	 *                           obsolete (i.e., the "non-obsolete" window is
+	 *                           exclusive).
+	 * @param agg                an aggregator to output a calculated "aggregated
+	 *                           signal strength"
+	 * @return a map from BSSID to an "aggregated wifiscan entry" (i.e., a single
+	 *         entry with its {@code signal} attribute modified to be the aggregated
+	 *         signal value instead of the value in just the most recent entry.
+	 */
+	public static Map<String, WifiScanEntry> getAggregatedWifiScans(Modeler.DataModel dataModel,
+			long obsoletionPeriodMs,
+			Aggregator<Double> agg) {
+		/*
+		 * NOTE: if a BSSID does not have an entry, it will not be returned
+		 * (i.e., it will not be a key in the returned map).
+		 */
+		Map<String, WifiScanEntry> aggregatedWifiScans = new HashMap<>();
+		for (Map.Entry<String, List<List<WifiScanEntry>>> mapEntry : dataModel.latestWifiScans.entrySet()) {
+			/*
+			 * Flatten the wifiscan entries and sort in reverse chronological order. Sorting
+			 * is done just in case the entries in the original list are not chronological
+			 * already - although they are inserted chronologically, perhaps latency,
+			 * synchronization, etc. could cause the actual unixTimeMs to be out-of-order.
+			 */
+			List<List<WifiScanEntry>> scans = mapEntry.getValue();
+			if (scans.isEmpty()) {
+				continue;
+			}
+			List<WifiScanEntry> mostRecentToOldest = scans.stream().flatMap(list -> list.stream())
+					.sorted((entry1, entry2) -> {
+						return -Long.compare(entry1.unixTimeMs, entry2.unixTimeMs);
+					}).collect(Collectors.toUnmodifiableList());
+
+			String bssid = mapEntry.getKey();
+			WifiScanEntry mostRecentEntry = mostRecentToOldest.get(0);
+			agg.reset();
+			for (WifiScanEntry entry : mostRecentToOldest) {
+				if (mostRecentEntry.unixTimeMs - entry.unixTimeMs >= obsoletionPeriodMs) {
```

Comment:
if I choose obsoletionPeriodMs as 0 (to indicate that I only want absolute recent), then, this loop will break and line 293 will be NullPointerException?

Reply:
Good point, perhaps making the non-obsolete window inclusive, meaning 0 ms => include the entry 0 ms ago. I'll try this.